File: CPP/Lesson_046_longest_substring_without_repeating_characters.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 046 -- Longest Substring Without Repeating Characters
// Category   : Sliding Window
// Difficulty : Medium
// Study Plan : Day 23
// =============================================================
//
// QUESTION:
//   Given a string, find the length of the longest substring without
//   repeating characters.
//
//   Example: "abcabcbb" -> 3
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> seen; int l=0, best=0;
        for (int r=0;r<(int)s.size();r++) {
            auto it = seen.find(s[r]);
            if (it!=seen.end() && it->second>=l) l = it->second+1;
            seen[s[r]] = r;
            best = max(best, r-l+1);
        }
        return best;
    }
};
int main(){ Solution s; cout<<s.lengthOfLongestSubstring("abcabcbb")<<" "<<s.lengthOfLongestSubstring("bbbbb")<<endl; }

File: Java/Lesson046_LongestSubstringWithoutRepeatingCharacters.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 046 -- Longest Substring Without Repeating Characters
// Category   : Sliding Window
// Difficulty : Medium
// Study Plan : Day 23
// =============================================================
//
// QUESTION:
//   Given a string, find the length of the longest substring without
//   repeating characters.
//
//   Example: "abcabcbb" -> 3
// =============================================================

import java.util.*;
public class Lesson046_LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer> seen = new HashMap<>();
        int l=0, best=0;
        for (int r=0;r<s.length();r++) {
            char c = s.charAt(r);
            if (seen.containsKey(c) && seen.get(c) >= l) l = seen.get(c)+1;
            seen.put(c, r);
            best = Math.max(best, r-l+1);
        }
        return best;
    }
    public static void main(String[] a){
        Lesson046_LongestSubstringWithoutRepeatingCharacters s = new Lesson046_LongestSubstringWithoutRepeatingCharacters();
        System.out.println(s.lengthOfLongestSubstring("abcabcbb"));
        System.out.println(s.lengthOfLongestSubstring("bbbbb"));
    }
}
